2017/proposals/bansal-utkarsh.md

```diff
@@ -0,0 +1,628 @@
+    ```python
+    class BaseWriterTest(object):
+      def __init__(self, reference):
+        self.ref = reference
+
+        # Instantiating a TempDir object
+        self.tmpdir = tempdir.TempDir()
+        self.reader = self.ref.reader(self.ref.trajectory)
+
+      def tmp_file(self, name):
+
+        # Note the usage to self.tmpdir here
+        return self.tmpdir.name + name + '.' + self.ref.ext
+
+    ```
+
+    Custom fixtures are easy to create. The tempdir instantiation part can be easily converted into a custom fixture as follows
+    ```python
+
+    # Creating a custom fixture
+    @pytest.fixture
+    def tmpdir():
+        return tempdir.TempDir()
+
+    class BaseWriterTest(object):
+      def __init__(self, reference):
+        self.ref = reference
+        # Note that we do not have to initialize self.tmpdir how
+        self.reader = self.ref.reader(self.ref.trajectory)
+
+      # Note that the method now takes an additional argument - tmpdir (our fixture)
+      def tmp_file(self, name, tmpdir):
+        # Note the usage to tmpdir here
+        return tmpdir.name + name + '.' + self.ref.ext
+    ```
+
+    Moreover fixtures have configurable scopes and can be used to cache resources that are reused, i.e. there is no need to initialize a particular resource everytime
+    it is needed in a different class/function. This will lead to performance gains too and will help reduce boilerplate code further.
+
+    ```python
+
+    # Note the usage of scope
+    @pytest.fixture(scope="function")
+    def custom_fixture():
+      return resource
+    ```
+
+    The scope can be "function", "class", "module", "session" or "invocation". This way we can cache resources that are being reused in other parts of the code.
+    As with the above example, many parts of the code require the `tmpdir` and as of now each class instantiates its own object. The scope of this fixture may be
+    changed to `"session"` so that the object is created only once.
+
+    Pytest also has some commonly used inbuilt fixtures.
+    An interesting one of them is `cache` that returns a cache object that can persist state between testing sessions.
+
+    ```python
+    cache.get(key, default)
+    cache.set(key, value)
+    ```
+
+    This fixture may be used to speed up tests and TravisCI builds in particular (which take very long time as of now - approx. 50 min.)
+
+2. Universal `assert` statements - pytest allows us to use the standard python assert for verifying expectations and values in tests. It also automagically provides
+quite helpful tracebacks when tests fail.
+
+    **Usage**
+
+    Nose has a number of `assert_*` methods such as `assert_equal`, `assert_raises`, etc. Pytest on the other hand uses only plain `assert` statements.
+    This is easier as developers don't have to remember all the methods and also pytest has awesome advanced introspection which brings great and
+    informative tracebacks when test cases fail.
```

Comment:
Never used sed but can look into it if needed. What I usually do is a search and replace in PyCharm and replace it on a case by case basis.